### PR TITLE
apollo-engine-reporting: direct path string building

### PR DESCRIPTION
**Pull request by pcarrier**
I don't have performance numbers, but creating those arrays and reversing them could be worth avoiding.

### CODE REVIEW

File: packages/apollo-engine-reporting/src/treeBuilder.ts [hunk 1]
```diff
@@ -233,14 +228,19 @@ function durationHrTimeToNanos(hrtime: [number, number]) {
 
 // Convert from the linked-list ResponsePath format to a dot-joined
 // string. Includes the full path (field names and array indices).
-function responsePathAsString(p: ResponsePath | undefined) {
+function responsePathAsString(p: ResponsePath | undefined): string {
```
Comment (abernix) [inline packages/apollo-engine-reporting/src/treeBuilder.ts]: I realize you didn't introduce it originally, but would you support making `p` optional and dropping the `| undefined`?

```suggestion
function responsePathAsString(p?: ResponsePath): string {
```

(and also leaving out the explicit `undefined` at the call site of this function on line 19, as I noted in a separate GH suggestion?)

File: packages/apollo-engine-reporting/src/treeBuilder.ts [hunk 2]
```diff
@@ -16,7 +11,7 @@ export class EngineReportingTreeBuilder {
   public startHrTime?: [number, number];
   private stopped = false;
   private nodes = new Map<string, Trace.Node>([
-    [rootResponsePath, this.rootNode],
+    [responsePathAsString(undefined), this.rootNode],
```
Comment (abernix) [inline packages/apollo-engine-reporting/src/treeBuilder.ts]: See comment below as to what this suggestion is suggesting.

```suggestion
    [responsePathAsString(), this.rootNode],
```

File: packages/apollo-engine-reporting/src/treeBuilder.ts [hunk 3]
```diff
@@ -233,14 +228,19 @@ function durationHrTimeToNanos(hrtime: [number, number]) {
...
-function responsePathAsString(p: ResponsePath | undefined) {
+function responsePathAsString(p: ResponsePath | undefined): string {
   if (p === undefined) {
     return '';
   }
-  return responsePathAsArray(p).join('.');
-}
 
-const rootResponsePath = responsePathAsString(undefined);
+  let res = String(p.key);
```
Comment (abernix) [inline packages/apollo-engine-reporting/src/treeBuilder.ts]: I'd love if we left a note to future adventurers as to why we took this approach! Thoughts on prefixing with, roughly:

```suggestion
  // A previous implementation used `responsePathAsArray` from `graphql-js/execution`,
  // however, that employed an approach that created new arrays unnecessarily.
  let res = String(p.key);
```

File: packages/apollo-engine-reporting/src/treeBuilder.ts [hunk 4]
```diff
@@ -1,10 +1,5 @@
-import {
-  GraphQLResolveInfo,
-  GraphQLError,
-  ResponsePath,
-  responsePathAsArray,
-} from 'graphql';
-import { Trace, google } from 'apollo-engine-reporting-protobuf';
+import { GraphQLError, GraphQLResolveInfo, ResponsePath } from 'graphql';
+import { google, Trace } from 'apollo-engine-reporting-protobuf';
```
Comment (abernix) [inline packages/apollo-engine-reporting/src/treeBuilder.ts]: Strongly prefer keeping these no-op, one-off changes out of change sets, but not asking you to revert it.

**Comment by abernix** [comment]
(I low-key wish there were `__tests__` for this `treeBuilder.ts` already, but alas, there are not.)

File: packages/apollo-engine-reporting/src/treeBuilder.ts (hunk 2 above)
Comment (pcarrier) [reply packages/apollo-engine-reporting/src/treeBuilder.ts]: I don't know the difference :)

**Pull request merged**